README.md
# action-release-download

GitHub Action: GitHub Release Download Action

## What this does, and does not do

This action will take the "assets_url" variable from a workflow triggered by a published release, and download all files based on the "browser_download_url" field.

## Usage

Here is an example on how to use this Action:

```
    - name: Release Download
      uses: sondreb/action-release-download@master
      with:
        token: ${{ secrets.GITHUB_TOKEN }}
        files: "action.js;README.md"
        folder: "${{github.workspace}}/package/"
        draft: true
        prerelease: true
        body: 'This is a pre-release'
        name: "Draft Release"
        tag: v0.0.1
```

### Notes

This is built as a quick and dirty proof-of-concept and is likely ridled with bugs and problems. Use at your own discretion.

## Releases

0.0.1

- Adds support for folders.

0.0.2

- Update packages and clean up test data and repo.

## License

[MIT](LICENSE)

CHANGELOG.md
## <small>0.0.2 (2022-01-06)</small>

* Add larger files for testing ([cec1088](https://github.com/sondreb/action-release-download/commit/cec1088))
* Add support for paging when there are many artifacts ([698fbc4](https://github.com/sondreb/action-release-download/commit/698fbc4))
* Bug fix ([5e9b216](https://github.com/sondreb/action-release-download/commit/5e9b216))
* Fix issue with link paging link not being provided by GitHub ([c40c068](https://github.com/sondreb/action-release-download/commit/c40c068))
* Improve the download process to respect redirects ([cf3efb1](https://github.com/sondreb/action-release-download/commit/cf3efb1))
* Remove the workflows ([90167f5](https://github.com/sondreb/action-release-download/commit/90167f5))
* Update packages and clean up repo ([9046298](https://github.com/sondreb/action-release-download/commit/9046298))



## <small>0.0.1 (2021-03-08)</small>

* Add a couple of nuget packages for testing ([0dd1d01](https://github.com/sondreb/action-release-download/commit/0dd1d01))
* Add complete download of artifact logic ([04d5144](https://github.com/sondreb/action-release-download/commit/04d5144))
* Add some extra logging to debug issues of download ([dc1106d](https://github.com/sondreb/action-release-download/commit/dc1106d))
* Add user agent to requests ([ff0e2a4](https://github.com/sondreb/action-release-download/commit/ff0e2a4))
* Add verbose logging ([4e4ec6d](https://github.com/sondreb/action-release-download/commit/4e4ec6d))
* Attempt to run using master and not main ([365ac77](https://github.com/sondreb/action-release-download/commit/365ac77))
* Begin development of artifact download action ([e8dc966](https://github.com/sondreb/action-release-download/commit/e8dc966))
* Cleanup of code ([ba72355](https://github.com/sondreb/action-release-download/commit/ba72355))
* Create folder if not exists ([cfe464b](https://github.com/sondreb/action-release-download/commit/cfe464b))
* Does the @main tag not work like @master? ([5fb8e0a](https://github.com/sondreb/action-release-download/commit/5fb8e0a))
* Getting there ([9156418](https://github.com/sondreb/action-release-download/commit/9156418))
* Initial commit ([ae60b05](https://github.com/sondreb/action-release-download/commit/ae60b05))
* It's getting late, syntax error. ([bf1f843](https://github.com/sondreb/action-release-download/commit/bf1f843))
* Make the build create a release ([d1a3063](https://github.com/sondreb/action-release-download/commit/d1a3063))
* Make the release use 'main' branch ([4ab0a6f](https://github.com/sondreb/action-release-download/commit/4ab0a6f))
* This is getting stupid? ([49d0b40](https://github.com/sondreb/action-release-download/commit/49d0b40))
* Update action.js ([88630ff](https://github.com/sondreb/action-release-download/commit/88630ff))
* Update action.js ([a165634](https://github.com/sondreb/action-release-download/commit/a165634))
* Update action.js ([e4d3c6c](https://github.com/sondreb/action-release-download/commit/e4d3c6c))
* Update action.js ([14409d1](https://github.com/sondreb/action-release-download/commit/14409d1))
* Update action.js ([cc77ac0](https://github.com/sondreb/action-release-download/commit/cc77ac0))
* Update action.js ([97efc43](https://github.com/sondreb/action-release-download/commit/97efc43))
* Update action.js ([ebc04bc](https://github.com/sondreb/action-release-download/commit/ebc04bc))
* Update branch to main ([b72d6cd](https://github.com/sondreb/action-release-download/commit/b72d6cd))
* Update release.yml ([399ed58](https://github.com/sondreb/action-release-download/commit/399ed58))
* Update to use the main branch on download action ([9aa6926](https://github.com/sondreb/action-release-download/commit/9aa6926))





test.js
const fs = require('fs');
const path = require('path');
// const core = require('@actions/core');
// const github = require('@actions/github');
const http = require('https');
const { parse } = require('url');
const { basename } = require('path');

let url = 'https://api.github.com/repositories/211867945/releases/39413272/assets';

const folder = './artifact';

const uri = parse(url);
const fileName = basename(uri.path);
const filePath = path.join(folder, fileName);
const file = fs.createWriteStream(filePath);

// const options = {
//     headers: { 'User-Agent': 'sondreb/action-release-download' }
// };

const options = {
    headers: { 'User-Agent': 'sondreb/action-release-download' }
};

function getFiles(urls, files, processLinks) {
    const url = urls.pop();

    if (!url) {
        return;
    }

    http.get(url, options, res => {
        let data = [];
        const headerDate = res.headers && res.headers.date ? res.headers.date : 'no response date';
        console.log('Status Code:', res.statusCode);
        console.log('Date in Response header:', headerDate);

        if (processLinks) {
            // Process all paging links, but only on first request.
            console.log(res.headers.link);
            var links = res.headers.link.split(', ');

            for (i = 0; i < links.length; i++) {
                let link = links[i];
                link = link.substring(1, link.indexOf('>'));
                console.log('link:' + link);
                urls.push(link);
            }
        }

        res.on('data', chunk => {
            data.push(chunk);
        });

        res.on('end', () => {
            console.log('Response ended: ');
            const assets = JSON.parse(Buffer.concat(data).toString());

            for (asset of assets) {
                console.log(`Download: ${asset.browser_download_url}`);
                files.push(asset.browser_download_url);
            }

            getFiles(urls, files, false);
            // downloadFiles(files);
        });
    }).on('error', err => {
        console.log('Error: ', err.message);
    });
}

const files = [];
let urls = ['https://api.github.com/repositories/211867945/releases/39413272/assets'];

getFiles(urls, files, true)


test-download.js
const fs = require('fs');
const path = require('path');
// const core = require('@actions/core');
// const github = require('@actions/github');
const http = require('https');
const { parse } = require('url');
const { basename } = require('path');

let url = 'https://github.com/sondreb/action-release-download/releases/download/v0.0.1/Blockcore.Features.BlockExplorer.1.0.6.nupkg';

const folder = './artifact';

const uri = parse(url);
const fileName = basename(uri.path);
const filePath = path.join(folder, fileName);

const file = fs.createWriteStream(filePath);

// const options = {
//     headers: { 'User-Agent': 'sondreb/action-release-download' }
// };

function get(url) {

    const options = {
        headers: { 'User-Agent': 'sondreb/action-release-download' }
    };

    http.get(url, options, (res) => {
        if (res.statusCode === 301 || res.statusCode === 302) {
            console.log('Forwarding to: ' + res.headers.location);
            return get(res.headers.location);
        }

        console.log('Saving file!');
        res.pipe(file);
    });
}

get(url);

// const request = http.get(url, options, function (response) {

//     if (res.statusCode === 301 || res.statusCode === 302) {
//         return get(res.headers.location, resolve, reject)
//     }

//     response.pipe(file);
// });

console.log('Done!');

// var parseChangelog = require('changelog-parser');

// parseChangelog(
//     {
//         filePath: 'CHANGELOG.md',
//         removeMarkdown: false
//     }
// )
//     .then(function (result) {
//         // changelog object
//         console.log(result)

//     })
//     .catch(function (err) {
//         console.error(err);
//     });



action.js
require('child_process')
    .execSync(
        'npm install @actions/core @actions/github conventional-changelog-cli mime',
        { cwd: __dirname }
    );

const fs = require('fs');
const path = require('path');
const mime = require('mime');
const core = require('@actions/core');
const github = require('@actions/github');
const http = require('https');
const { parse } = require('url');
const { basename } = require('path');
const { callbackify } = require('util');

(async () => {
    try {
        // const api = github.getOctokit(core.getInput('token'));
        const url = core.getInput('url');

        // Create an array of the initial URL, which will be populated with paging URLs after initial request.
        let urls = [url];

        const verbose = core.getInput('verbose') == 'true'; // input is always string, not boolean.
        const folder = core.getInput('folder');
        var timeout = 10000;

        // Options for HTTP requests against GitHub APIs. User-Agent is required.
        const options = {
            headers: { 'User-Agent': 'sondreb/action-release-download' }
        };

        info(`Url input: ${url}`);
        info(`Folder input: ${folder}`);

        if (!fs.existsSync(folder)) {
            info(`Creating the destination folder: "${folder}".`);
            fs.mkdirSync(folder);
        }

        function info(text, ...params) {
            log(text, true, ...params);
        }

        function debug(text, ...params) {
            log(text, verbose, ...params);
        }

        function log(text, enabled, ...params) {
            if (enabled) {
                console.log(text, ...params);
            }
        }

        var timeout_wrapper = function (req) {
            return function () {
                console.log('abort');
                req.abort();
                // callback(true, { size: 0, downloaded: 0, progress: 0, status: 'Timeout' }, "File transfer timeout!");
            };
        };

        function completed() {
            info('All is norminal 🚀. Execution has ended.');
        }

        function get(url, file, callback) {

            const options = {
                headers: { 'User-Agent': 'sondreb/action-release-download' }
            };

            http.get(url, options, (res) => {
                if (res.statusCode === 301 || res.statusCode === 302) {
                    console.log('Forwarding to: ' + res.headers.location);
                    return get(res.headers.location, file, callback);
                }

                debug(`Download completed: ${url}`);

                res.pipe(file);
                callback();
            });
        }

        function downloadFiles(urls) {
            var url = urls.pop();

            if (!url) {
                completed();
                return;
            }

            const uri = parse(url);
            const fileName = basename(uri.path);
            const filePath = path.join(folder, fileName);

            var file = fs.createWriteStream(filePath);

            // Get the file then continue on the next.
            get(url, file, () => {
                downloadFiles(urls);
            });
        }

        function getFiles(urls, files, processLinks) {
            const url = urls.pop();

            // When all URLs has been processes, start downloading the files.
            if (!url) {
                downloadFiles(files);
                return;
            }

            http.get(url, options, res => {
                let data = [];
                const headerDate = res.headers && res.headers.date ? res.headers.date : 'no response date';
                console.log('Status Code:', res.statusCode);
                console.log('Date in Response header:', headerDate);

                if (processLinks) {
                    // Process all paging links, but only on first request.
                    if (res.headers.link) {
                        var links = res.headers.link.split(', ');

                        for (i = 0; i < links.length; i++) {
                            let link = links[i];
                            link = link.substring(1, link.indexOf('>'));
                            console.log('link:' + link);
                            urls.push(link);
                        }
                    }
                }

                res.on('data', chunk => {
                    data.push(chunk);
                });

                res.on('end', () => {
                    console.log('Response ended: ');
                    const assets = JSON.parse(Buffer.concat(data).toString());

                    for (asset of assets) {
                        console.log(`Download: ${asset.browser_download_url}`);
                        files.push(asset.browser_download_url);
                    }

                    getFiles(urls, files, false);
                });
            }).on('error', err => {
                console.log('Error: ', err.message);
            });
        }

        const files = [];

        getFiles(urls, files, true);

    } catch (error) {
        console.error(error);
        core.setFailed(error.message);
    }
})();